### PR TITLE
Feature/reuse dot1x realms integration

### LINKED ISSUE #1543: reuse_dot1x_credentials: security concern
When using portal profile 'reuse_dot1x_credentials' option along with 'stripped_username' in a source, it may cause a security concern in the sense that:
- Connecting the a EAP SSID with valid credentials
- Successfully mschap with full credentials (username@realm)
- On the portal, using 'reuse_dot1x_credentials'
- Stripping the realm part
- SECURITY CONCERN: May match the same username from a different source with different rights since there is no bind

Solution:
Enforce the use of sources with configured realms when using 'reuse_dot1x_credentials' option.
- One can only match a source with the same realm
- One can match the local source configured with a local realm

### PR DESCRIPTION
**Pull request by dwlfrth**
# Description

FreeRADIUS "default" (DEFAULT, NULL, LOCAL) realms are now handled by PacketFence realm feature.
reuseDot1xCredentials feature only work for the provided realm with the configured authentication source
# Impacts

Portal (reuseDot1xCredentials flow)
Realms
FreeRADIUS realms
# Issue

fixes #1543 
# Delete branch after merge

YES
# NEWS file entries
## Bug Fixes
- Address reuseDot1xCredentials feature security concern to match different authentication source than the one authenticated with (#1543)
# UPGRADE file entries

reuseDot1xCredentials portal profile feature now requires to have configured realms. The realms also need to have authentication sources configured otherwise none will be matched. As a security precaution, this option will only reuse dot1x if there is an authentication source matching the provided realm. This means, if users use 802.1x credentials with a domain part (username@domain, domain\username), the domain part need to be configured as a realm under the RADIUS section and an authentication source need to be configured for that realm. If users do not use 802.1x credentials with a domain part, only the NULL realm will be match IF an authentication source is configured for it.

FreeRADIUS default realms (DEFAULT, NULL, LOCAL) are now handled by the PacketFence realm feature. Make sure to migrate any existing custom configuration for theses realms (if any) to the PacketFence configuration.

### CODE REVIEW

**Comment by dwlfrth** [comment]
"review" me

**Comment by julsemaan** [comment]
Would be nice if you integrated the logic definition that is in the upgrade notes in the administration guide.

File: html/captive-portal/lib/captiveportal/PacketFence/DynamicRouting/Module/Authentication/Login.pm [hunk 1]
```diff
@@ -145,30 +145,49 @@ sub authenticate {
         return;
     }
 
-    if(isenabled($self->app->profile->reuseDot1xCredentials)) {
+    if ( isenabled($self->app->profile->reuseDot1xCredentials) ) {
         my $mac       = $self->current_mac;
         my $node_info = node_view($mac);
         ($username,$realm) = strip_username($node_info->{'last_dot1x_username'});
-        get_logger->info("Reusing 802.1x credentials. Gave username ; $username");
+        $realm = "default" unless(defined($realm));
```
Comment (julsemaan) [inline html/captive-portal/lib/captiveportal/PacketFence/DynamicRouting/Module/Authentication/Login.pm]: Gimme some of them constants

File: html/pfappserver/lib/pfappserver/Form/Config/ProfileCommon.pm [hunk 2]
```diff
@@ -283,6 +283,10 @@ has_field 'reuse_dot1x_credentials' =>
     type => 'Checkbox',
     checkbox_value => 'enabled',
     unchecked_value => 'disabled',
+    tags => {
+        after_element   =>    \&help,
+        help            => 'This option kind of act as a SSO when someone needs to face the captive portal after a successful 802.1x connection. 802.1x credentials are reused on the portal to match an authentication and get the appropriate actions. As a security precaution, this option will only reuse 802.1x credentials if there is an authentication source matching the provided realm. This means, if users use 802.1x credentials with a domain part (username@domain, domain\username), the domain part need to be configured as a realm under the RADIUS section and an authentication source need to be configured for that realm. If users do not use 802.1x credentials with a domain part, only the DEFAULT realm will be match IF an authentication source is configured for it.'
```
Comment (julsemaan) [inline html/pfappserver/lib/pfappserver/Form/Config/ProfileCommon.pm]: need -> needs everywhere in this

Comment (julsemaan) [inline html/pfappserver/lib/pfappserver/Form/Config/ProfileCommon.pm]: 'kind of act' sounds a bit 'unprofessional'

Maybe 'This option emulates SSO' ?

**Comment by dwlfrth** [comment]
review me

File: docs/PacketFence_Administration_Guide.asciidoc [hunk 3]
```diff
@@ -850,6 +850,13 @@ The other files in this directory are managed by PacketFence using templates, so
 
 Upon PacketFence installation, self-signed certificates will be created in `/usr/local/pf/conf/ssl` (`server.key` and `server.crt`). Those certificates can be replaced anytime by your 3rd-party or existing wildcard certificate without problems. Please note that the CN (Common Name) needs to be the same as the one defined in the PacketFence configuration file (`pf.conf`).
 
+Reuse 802.1x credentials
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Under certain circumstances, like per example displays an AUP after a successful 802.1x connection, it might be interesting to have the ability to use a "SSO emulation" in the sense that the user do not need to reenter credentials on the portal after having entered them during the 802.1x EAP process. The 'Reuse 802.1x credentials' portal profile option perfectly address this purpose. The same user as the one used during the 802.1x connection will be used against the different portal profile authentication source to recompute the role from the portal.
```
Comment (julsemaan) [inline docs/PacketFence_Administration_Guide.asciidoc]: Under certain circumstances, for example to show an AUP...

Comment (julsemaan) [inline docs/PacketFence_Administration_Guide.asciidoc]: use an "SSO emulation"

Comment (julsemaan) [inline docs/PacketFence_Administration_Guide.asciidoc]: that the user *doesn't need to *re-enter his credentials

Comment (julsemaan) [inline docs/PacketFence_Administration_Guide.asciidoc]: option *will address this purpose

Comment (julsemaan) [inline docs/PacketFence_Administration_Guide.asciidoc]: The same *username

Comment (julsemaan) [inline docs/PacketFence_Administration_Guide.asciidoc]: *authentication sources

File: docs/PacketFence_Administration_Guide.asciidoc [hunk 4]
```diff
@@ -850,6 +850,13 @@ The other files in this directory are managed by PacketFence using templates, so
...
+
+As a security precaution, this option will only reuse 802.1x credentials if there is an authentication source matching the provided realm. This means, if users use 802.1x credentials with a domain part (username@domain, domain\username), the domain part needs to be configured as a realm under the RADIUS section and an authentication source needs to be configured for that realm. If users do not use 802.1x credentials with a domain part, only the DEFAULT realm will be match IF an authentication source is configured for it.
```
Comment (julsemaan) [inline docs/PacketFence_Administration_Guide.asciidoc]: if they do not put a domain, it should be NULL, not default that is used...
I guess that needs fixing in the code too

**Comment by dwlfrth** [comment]
Review me

File: html/captive-portal/lib/captiveportal/PacketFence/DynamicRouting/Module/Authentication/Login.pm [hunk 5]
```diff
@@ -145,30 +146,49 @@ sub authenticate {
...
+
+        # Fetch appropriate source to use with 'reuseDot1xCredentials' feature
+        my $source = pf::config::util::get_realm_source($username, $realm);
+        
+        # No source found for specified realm
+        unless ( defined($source) ) {
+            get_logger->error("Unable to find an authentication source for the specified realm '$realm' while using reuseDot1xCredentials");
+            $self->app->flash->{error} = "Cannot find a valid authentication source for '" . $node_info->{'last_dot1x_username'} . "'";
+
+            $self->done();
```
Comment (julsemaan) [inline html/captive-portal/lib/captiveportal/PacketFence/DynamicRouting/Module/Authentication/Login.pm]: Done actually marks this module as complete and continues the flow.

`done != detach`

Not sure that is really what you want

File: html/captive-portal/lib/captiveportal/PacketFence/DynamicRouting/Module/Authentication/Login.pm [hunk 6]
```diff
@@ -145,30 +146,49 @@ sub authenticate {
...
-        } else {
-            get_logger->error("Reusing 802.1x credentials but not able to find the source for username $username");
-            $self->app->flash->{error} = "Reusing 802.1x credentials but not able to find the source for username $username, did you define the source on the portal ?";
         }
-    } else {
+        else {
+            get_logger->error("Unable to find a match in the '$realm' realm authentication source for credentials '" . $node_info->{'last_dot1x_username'} . "' while using reuseDot1xCredentials");
+            $self->app->flash->{error} = "Cannot find a valid authentication source for '" . $node_info->{'last_dot1x_username'} . "'";
+
+            $self->done();
```
Comment (julsemaan) [inline html/captive-portal/lib/captiveportal/PacketFence/DynamicRouting/Module/Authentication/Login.pm]: Same here, don't think done should be called

**Comment by dwlfrth** [comment]
Review me

**Pull request merged**